ui/src/app/namespaces/namespacesSlice.ts

```diff
@@ -0,0 +1,128 @@
+/* eslint-disable @typescript-eslint/no-use-before-define */
+import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
+import {
+  createNamespace,
+  deleteNamespace,
+  listNamespaces,
+  updateNamespace
+} from '~/data/api';
+import { RootState } from '~/store';
+import { LoadingStatus } from '~/types/Meta';
+import { INamespace, INamespaceBase } from '~/types/Namespace';
+
+interface INamespacesState {
+  namespaces: { [key: string]: INamespace };
+  status: LoadingStatus;
+  currentNamespace: string;
+  error: string | undefined;
+}
+
+const initialState: INamespacesState = {
+  namespaces: {},
+  status: LoadingStatus.IDLE,
+  currentNamespace: 'default',
+  error: undefined
+};
+
+export const namespacesSlice = createSlice({
+  name: 'namespaces',
+  initialState,
+  reducers: {
+    namespaceCreated: (state, action) => {
+      const namespace = action.payload;
+      state.namespaces[namespace.key] = action.payload;
+    },
+    namespaceUpdated: (state, action) => {
+      const namespace = action.payload;
+      state.namespaces[namespace.key] = action.payload;
+    },
+    namespaceDeleted: (state, action) => {
+      const namespace = action.payload;
+      delete state.namespaces[namespace.key];
+    },
+    currentNamespaceChanged: (state, action) => {
+      const namespace = action.payload;
+      state.currentNamespace = namespace.key;
+    }
+  },
+  extraReducers(builder) {
+    builder
+      .addCase(fetchNamespacesAsync.pending, (state, _action) => {
+        state.status = LoadingStatus.LOADING;
+      })
+      .addCase(fetchNamespacesAsync.fulfilled, (state, action) => {
+        state.status = LoadingStatus.SUCCEEDED;
+        state.namespaces = action.payload;
+      })
+      .addCase(fetchNamespacesAsync.rejected, (state, action) => {
+        state.status = LoadingStatus.FAILED;
+        state.error = action.error.message;
+      })
+      .addCase(createNamespaceAsync.fulfilled, (state, action) => {
+        const namespace = action.payload;
+        state.namespaces[namespace.key] = namespace;
+      })
+      .addCase(updateNamespaceAsync.fulfilled, (state, action) => {
+        const namespace = action.payload;
+        state.namespaces[namespace.key] = namespace;
+      })
+      .addCase(deleteNamespaceAsync.fulfilled, (state, action) => {
+        const key = action.payload;
+        delete state.namespaces[key];
+      });
+  }
+});
+
+export const { namespaceCreated, currentNamespaceChanged } =
+  namespacesSlice.actions;
+
+export const selectNamespaces = (state: RootState) =>
+  Object.entries(state.namespaces.namespaces).map(
+    ([_, value]) => value
+  ) as INamespace[];
+
+export const selectCurrentNamespace = (state: RootState) => {
+  if (state.namespaces.status === LoadingStatus.SUCCEEDED) {
+    return state.namespaces.namespaces[state.namespaces.currentNamespace];
+  }
+  return { key: '', name: '', description: '' } as INamespace;
```

Comment:
Would it be safe to return the following here (genuine question):
```suggestion
  return { key: 'default', name: 'Default', description: '' } as INamespace;
```

Reply:
yah we can do that. it should only be shown for a split second in the listbox until all the namespaces load from the fetch function on initial app load